subsys/jwt/jwt.c

```diff
@@ -190,120 +180,27 @@ int jwt_add_payload(struct jwt_builder *builder,
 	return res;
 }
 
-#ifdef CONFIG_JWT_SIGN_RSA
-
-static int csprng_wrapper(void *ctx, unsigned char *dest, size_t size)
-{
-	ARG_UNUSED(ctx);
-
-	return sys_csrand_get((void *)dest, size);
-}
-
-int jwt_sign(struct jwt_builder *builder,
-	     const char *der_key,
-	     size_t der_key_len)
-{
-	int res;
-	mbedtls_pk_context ctx;
-
-	mbedtls_pk_init(&ctx);
-
-	res = mbedtls_pk_parse_key(&ctx, der_key, der_key_len,
-			       NULL, 0, csprng_wrapper, NULL);
-	if (res != 0) {
-		return res;
-	}
-
-	uint8_t hash[32], sig[256];
-	size_t sig_len = sizeof(sig);
-
-	/*
-	 * The '0' indicates to mbedtls to do a SHA256, instead of
-	 * 224.
-	 */
-	mbedtls_sha256(builder->base, builder->buf - builder->base,
-		       hash, 0);
-
-	res = mbedtls_pk_sign(&ctx, MBEDTLS_MD_SHA256,
-			      hash, sizeof(hash),
-			      sig, sig_len, &sig_len,
-			      csprng_wrapper, NULL);
-	if (res != 0) {
-		return res;
-	}
-
-	base64_outch(builder, '.');
-	base64_append_bytes(sig, sig_len, builder);
-	base64_flush(builder);
-
-	return builder->overflowed ? -ENOMEM : 0;
-}
-#endif
-
-#ifdef CONFIG_JWT_SIGN_ECDSA
-static TCCtrPrng_t prng_state;
-static bool prng_init;
-
-static const char personalize[] = "zephyr:drivers/jwt/jwt.c";
-
-static int setup_prng(void)
-{
-	if (prng_init) {
-		return 0;
-	}
-	prng_init = true;
-
-	uint8_t entropy[TC_AES_KEY_SIZE + TC_AES_BLOCK_SIZE];
-
-	sys_rand_get(entropy, sizeof(entropy));
-
-	int res = tc_ctr_prng_init(&prng_state,
-				   (const uint8_t *) &entropy, sizeof(entropy),
-				   personalize,
-				   sizeof(personalize));
-
-	return res == TC_CRYPTO_SUCCESS ? 0 : -EINVAL;
-}
-
-int default_CSPRNG(uint8_t *dest, unsigned int size)
-{
-	int res = tc_ctr_prng_generate(&prng_state, NULL, 0, dest, size);
-	return res;
-}
+int jwt_sign_impl(struct jwt_builder *builder, const unsigned char *der_key,
+		  size_t der_key_len, unsigned char *sig, size_t sig_size);
 
 int jwt_sign(struct jwt_builder *builder,
 	     const char *der_key,
 	     size_t der_key_len)
 {
-	struct tc_sha256_state_struct ctx;
-	uint8_t hash[32], sig[64];
-	int res;
-
-	tc_sha256_init(&ctx);
-	tc_sha256_update(&ctx, builder->base, builder->buf - builder->base);
-	tc_sha256_final(hash, &ctx);
+	int ret;
+	unsigned char sig[JWT_SIGNATURE_LEN];
 
-	res = setup_prng();
-
-	if (res != 0) {
-		return res;
-	}
-	uECC_set_rng(&default_CSPRNG);
```

Comment:
This call was not needed?

Reply:
Yeah, it's a bit tricky. `uECC_set_rng` allows you to define a custom RNG function, but if you check in `modules/crypto/tinycrypt/lib/source/ecc.c` you'll see that:
```
#if default_RNG_defined
static uECC_RNG_Function g_rng_function = &default_CSPRNG;
#else
static uECC_RNG_Function g_rng_function = 0;
#endif
```
and `default_RNG_defined` is always set (defined in `modules/crypto/tinycrypt/lib/include/tinycrypt/ecc_platform_specific.h`).
So that line that I removed just sets the very same function that is already set by default in TinyCrypt. Not really useful